File: 06/sol.py
def read_file(filename="input.txt"):
    with open(filename, "r") as f:
        return f.read().rstrip()


def unique(s):
    return len(set(s)) == len(s)


def solve1(s):
    for i in range(4, len(s)+1):
        if unique(s[i-4:i]):
            return i
    return None


def solve2(s):
    for i in range(14, len(s)+1):
        if unique(s[i-14:i]):
            return i
    return None


if __name__ == "__main__":
    input_string = read_file()
    assert input_string is not None
    res1 = solve1(input_string)
    print(res1)
    res2 = solve2(input_string)
    print(res2)

File: 06/sol.jl
function readstring(filename)
    input = rstrip(read(filename, String))
    @assert !isempty(input) "File Empty"
    return input
end

function parseinput(input)
    return strip(input)
end

function unique(s::AbstractString)::Bool
    return length(Set(s)) == length(s)
end

function marker(n::Int, s::AbstractString)::Int
    for i in 1:(length(s)-n+1)
        if unique(s[i:i+n-1])
            return i+n-1
        end
    end
end

function solve1(x)
    return marker(4, x)
end

function solve2(x)
    return marker(14, x)
end

if abspath(PROGRAM_FILE) == @__FILE__
    filename = isempty(ARGS) ? "input.txt" : ARGS[1]
    input = readstring(filename)
    x = parseinput(input)
    r1 = solve1(x)
    println(r1)
    r2 = solve2(x)
    println(r2)
end
